vm/src/stdlib/itertools.rs

```diff
@@ -650,6 +654,114 @@ impl PyItertoolsAccumulate {
     }
 }
 
+#[derive(Debug)]
+struct PyItertoolsTeeData {
+    iterable: PyObjectRef,
+    values: RefCell<Vec<PyObjectRef>>,
+}
+
+impl PyItertoolsTeeData {
+    fn new(
+        iterable: PyObjectRef,
+        vm: &VirtualMachine,
+    ) -> Result<Rc<PyItertoolsTeeData>, PyObjectRef> {
+        Ok(Rc::new(PyItertoolsTeeData {
+            iterable: get_iter(vm, &iterable)?,
+            values: RefCell::new(vec![]),
+        }))
+    }
+
+    fn get_item(&self, vm: &VirtualMachine, index: usize) -> PyResult {
+        if self.values.borrow().len() == index {
+            let result = call_next(vm, &self.iterable)?;
+            self.values.borrow_mut().push(result);
+        }
+        Ok(self.values.borrow()[index].clone())
+    }
+}
+
+#[pyclass]
+#[derive(Debug)]
+struct PyItertoolsTee {
+    tee_data: Rc<PyItertoolsTeeData>,
+    index: Cell<usize>,
+}
+
+impl PyValue for PyItertoolsTee {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("itertools", "tee")
+    }
+}
+
+#[pyimpl]
+impl PyItertoolsTee {
+    fn from_iter(iterable: PyObjectRef, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
+        let it = get_iter(vm, &iterable)?;
+        if it.class().is(&PyItertoolsTee::class(vm)) {
+            return vm.call_method(&it, "__copy__", PyFuncArgs::from(vec![]));
+        }
+        Ok(PyItertoolsTee {
+            tee_data: PyItertoolsTeeData::new(it, vm)?,
+            index: Cell::from(0),
+        }
+        .into_ref_with_type(vm, PyItertoolsTee::class(vm))?
+        .into_object())
+    }
+
+    #[pymethod(name = "__new__")]
+    #[allow(clippy::new_ret_no_self)]
+    fn new(
+        _cls: PyClassRef,
+        iterable: PyObjectRef,
+        n: OptionalArg<PyIntRef>,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyRef<PyTuple>> {
+        let n = match n {
+            OptionalArg::Present(x) => match x.as_bigint().to_usize() {
+                Some(y) => y,
+                None => return Err(vm.new_overflow_error(String::from("n is too big"))),
+            },
+            OptionalArg::Missing => 2,
+        };
+
+        let copyable = if objtype::class_has_attr(&iterable.class(), "__copy__") {
+            vm.call_method(&iterable, "__copy__", PyFuncArgs::from(vec![]))?
```

Comment:
It seems this line is not tested by the new snippets tests. Adding one will be helpful for review